### DISCUSSION TITLE
GitHub Repo and/or Project structure

### DISCUSSION THREAD

**Discussion by BilldarBagdar** [markusschloesser/MackieC4_P3#17] [Ideas]
Some things I've noticed since starting to mess around.  

I suggest introducing a bit of a folder structure in the "repository".  Currently, this repository is completely flat, all the files are the "Midi Remote Scripts/MackieC4/" remote script files except one, the README.txt is a "git specific file".  And there are other git specific files such as .gitignore I suggest we add to the repo that don't really belong to the actual remote script.  

![projectStructure](https://user-images.githubusercontent.com/80170376/111206131-fd3a2680-8595-11eb-9d2a-2a84ea6a147b.PNG)  This shows my "personal repo" from last spring next to your "public repo" in the same local Pycharm project as an example of the kind of structure I'm recomending for the repo.

- I suggest setting up a "release" folder under which we would maintain a MackieC4 folder that would in turn contain the code files we have "released" so the MackieC4 folder and  "release" contents are simply ready for copy and paste by anyone into their Live remote scripts folder. (Even before there actually is any "code to release", setting up an empty release area represents the goal.)  
-- Then we also set up a wip folder that contains another MackieC4 folder that in turn contains all the "working files" where even the master branch might entertain more frequent updates than the "release" version.  Basically all the files in the current repo except README would slide down under a [repo root]/wip/MackieC4 folder.  That way, the same easy copy/paste of the MackieC4 folder (and wip contents) over to Live is available. (Files can be copy/pasted many ways, the folder structure in the repository is the point of the example)
--Then thirdly, we leave the repository "root level" for git specific stuff like the README and .gitignore files, maybe other folders for related files.  One of the things we can add to the ignore file is the PyCharm project files (ignore everything under <Pycharm project root>/.idea/ for example) which in my experience are best not to share since the items under .idea/ tend to contain lots of "local machine specific" paths.  And also, it's possible someone else could want to contribute and who doesn't use PyCharm and all those files would get in that theoretical person's way.)

I was able to make all the updates I talked about in a branch called folders, and "pushed it" so you should be able to "fetch it" from origin and then "checkout folders" to have a look around locally (or even look around at the folders branch here on GitHub).  But I didn't open a Pull request, I think that would be a little presumptuous at this point, having just sprung the idea on you.  If you look around and you like it, or don't mind, Pull it on in.  If not, tell me what you don't like about it and hopefully we can find a way forward.  When I reopened Pycharm after making the "git repository folder structure" changes, all those changes just appeared in PyCharm.  But you could need to close and reopen files from new locations.

If you want any details about what commands I ran in git to make these changes in the branch, or maybe what it means when I use git-words like origin, checkout, and fetch, just ask.  The more you surprise me with what you already know, the less I want to get caught flat footed another time.

**Comment by markusschloesser** [comment]
makes total sense, all of it.
I started with an experimental branch yesterday and got confused for an hour our so, because I couldn't figure out where pycharm stored those files locally. I thought there should have been another local folder 📂 for that branch, but if I now understand correctly, local stuff is a replica of the checked- out GH stuff and there is only ONE local folder? I'm ok with that, but took me a minute 😂.

I need to gain more experience about file handling with pycharm and GH because I'm old school and usually do stuff on a file level (only learned about "commit and push" last week, frankly "commit and upload" would have been a better wording). And I don't wanna fuck things up accidentally.
I understand most stuff now, but haven't taught myself about "rebase" and a couple of other things, but pycharm's help pages are fine for me :-) Some stuff is also a little bit harder for me cos I'm not a native speaker. Thanks for the offer though! I might take you up on it at some point.

**Reply by markusschloesser** [reply]
btw how about a "tools" folder? I have that locally and store stuff like uncompyle/decompile and the stub building stuff plus xml in there. Do you want that?

**Comment by markusschloesser** [comment]
Ok I now have a MS_stable and a MS_experimental branch and will use those two for my stuff here.
1. How do I / you move those branches into the folder structure? With pyvharm?
2. Do you use something similar? stable/exp
3. How do we distinguish between Live11/py3 and Live10/py2?

**Comment by BilldarBagdar** [comment]
Pycharm has pretty good integration with "source control" whatever the flavor, but under the hood, git always has more powerful features that an IDE can make difficult just because the IDE is focused elsewhere (source control represents a kind of rabbit hole compared to an IDE's "core mission").  That's why I recommended "Sourcetree" as a way to understand your "git repository".  It comes bundled with a version of the "git bash" command line tool, and that command line is where I do most of my "git work".  All I do in PyCharm is "commits".  Then I go to the git-bash command line and run "actual git commands".

Yes, there **is** only one copy of any file (in a git repository) in your local file system, and it will be the version of the file in the "checked out branch".  But you can always update the file system to any branch you want just by checking out a new or existing branch.  Git won't let you "destroy local changes" when switching branches.  So there are times when you need to "roll back" or "stash" changes to the current branch that would be overwritten by switching to an existing branch.

I also tend to continue "reusing" existing branches with generic names like "experimental".  But git is built around branches and branch handling, and we lose a lot of "development history" by not "properly using" the branching mechanic.  The basic premise is that a branch represents some "feature" a developer wants to implement.  For example, the idea is when you sit down to tackle the "exploding log file" problem, you also start a branch named "fixodu" (for: fix on_display_update()) or whatever.  If you solve the problem, you "commit and push" the branch to "origin" and then "Pull" the branch (the changes) into master, and then delete the Branch (because it is now a "copy of master" anyway) and check that item off the "punch list".  If you don't solve the problem, then you either keep working on it or you just delete the branch without committing (to GitHub) or pulling anything (into master branch).

3.  I don't know yet.  But that is the first thing I am working on.  I get an immediate error and the script doesn't "load properly" when Live 10 tries loading my "MarkusC4test" script (of your Py3 Live 11 code).  The problem logged is "no module named builtins" (line 6 of `__init()__`).  I think the problem is something I can fix, and then I suspect we would need to figure out how to represent both the differences and the similarities between the Live 10 (legacy) version and the Live 11+ version within the repository in a way that is easy for people to figure out. (and easy for us to maintain)

I would have added a "utilities" folder already, but I didn't want to change too much structure initially before you were even introduced to the idea.  The only comment I would have about a "tools" folder is whatever "tools" or "utilities" are in there should be "of general use" to this "repository community".  If you have something that is great for you, but would need to be almost completely re-written to be any good for anyone else, I would suggest that is NOT something of general use to the community.  But for example, I wrote a PowerShell script that could be generally useful to other developers after a few minor updates like the path to your repo on your file system.  But, for example, if that PowerShell script depended on some third party tool (that isn't freely available) to do what it does, then that PowerShell script would be much less of a candidate for inclusion in a "tools" or "utilities" folder.

and we don't have to stop at one additional folder, or make any permanent decisions (at this point).  This is our project and it should be organized exactly as we see fit.  The way it makes the most sense to us.

**Reply by markusschloesser** [reply]
re 3. what I've seen elsewhere is:
```
import sys
if sys.version_info[0] >= 3:
    import something
else:
    import something else
```
and yes `builtins `gets a "not available in py2" error in pycharm

**Reply by BilldarBagdar** [reply]
Yes, that kind of configuration guard helps with the Py3 versus Py2.7 stuff and/but I still need to deal with any Live 10 versus Live 11 differences.  If that `Live.MidiMap.map_midi_cc_with_feedback_map()` method signature changed between 10 and 11, for example, there could be lots of other related changes...

**Reply by markusschloesser** [reply]
> Pycharm has pretty good integration with "source control" whatever the flavor, but under the hood, git always has more powerful features that an IDE can make difficult just because the IDE is focused elsewhere (source control represents a kind of rabbit hole compared to an IDE's "core mission"). That's why I recommended "Sourcetree" as a way to understand your "git repository". It comes bundled with a version of the "git bash" command line tool, and that command line is where I do most of my "git work". All I do in PyCharm is "commits". Then I go to the git-bash command line and run "actual git commands".

MS that's currently a bit too much overhead for me, I'll probably need another week or two for that ;-) In the meantime, I'll use pycharm's git.


> 
> Yes, there **is** only one copy of any file (in a git repository) in your local file system, and it will be the version of the file in the "checked out branch". But you can always update the file system to any branch you want just by checking out a new or existing branch. Git won't let you "destroy local changes" when switching branches. So there are times when you need to "roll back" or "stash" changes to the current branch that would be overwritten by switching to an existing branch.

MS: yep, it "happened" to me already, seems to be fine for now, the only thing I need to stop doing, it working with "local history" in pycharm, it keeps messing up things cos it doesn't take branches into consideration. So I'll be using compare to gh branches a lot more from now on.

> 
> I also tend to continue "reusing" existing branches with generic names like "experimental". But git is built around branches and branch handling, and we lose a lot of "development history" by not "properly using" the branching mechanic. The basic premise is that a branch represents some "feature" a developer wants to implement. For example, the idea is when you sit down to tackle the "exploding log file" problem, you also start a branch named "fixodu" (for: fix on_display_update()) or whatever. If you solve the problem, you "commit and push" the branch to "origin" and then "Pull" the branch (the changes) into master, and then delete the Branch (because it is now a "copy of master" anyway) and check that item off the "punch list". If you don't solve the problem, then you either keep working on it or you just delete the branch without committing (to GitHub) or pulling anything (into master branch).

MS: got it, thanks!



> I would have added a "utilities" folder already, but I didn't want to change too much structure initially before you were even introduced to the idea. The only comment I would have about a "tools" folder is whatever "tools" or "utilities" are in there should be "of general use" to this "repository community". If you have something that is great for you, but would need to be almost completely re-written to be any good for anyone else, I would suggest that is NOT something of general use to the community. But for example, I wrote a PowerShell script that could be generally useful to other developers after a few minor updates like the path to your repo on your file system. But, for example, if that PowerShell script depended on some third party tool (that isn't freely available) to do what it does, then that PowerShell script would be much less of a candidate for inclusion in a "tools" or "utilities" folder.
> 
> and we don't have to stop at one additional folder, or make any permanent decisions (at this point). This is our project and it should be organized exactly as we see fit. The way it makes the most sense to us.

MS: OK, let's add folders! :-)

**Reply by markusschloesser** [reply]
> Yes, that kind of configuration guard helps with the Py3 versus Py2.7 stuff and/but I still need to deal with any Live 10 versus Live 11 differences. If that `Live.MidiMap.map_midi_cc_with_feedback_map()` method signature changed between 10 and 11, for example, there could be lots of other related changes...

You/ we can check for Live version with:
````
import Live 

app = Live.Application.get_application()
IS_LIVE_10 = app.get_major_version() == 10
IS_LIVE_9 = app.get_major_version() >= 9
IS_LIVE_9_1 = IS_LIVE_10 or (IS_LIVE_9 and app.get_minor_version() >= 1)
IS_LIVE_9_5 = IS_LIVE_10 or (IS_LIVE_9 and app.get_minor_version() >= 5)
````
and that use that for distinguishing where it needs to be done. But I'd strongly suggest to only use Live 10 and 11.
Found that here: https://github.com/ldrolez/clyphx-live10/blob/master/ClyphX/consts.py

**Comment by BilldarBagdar** [comment]
For example, when I created the "release" and "wip" folders, I did it in the file system like normal.  But then I had to issue a git command to tell git the changes are legitimate and should be tracked as "changes to the repository branch" (a command like: `git add release/MackieC4/releasePlaceholder.py` will add the folders to the repo.  Git doesn't store "empty folders" in repositories, only files. and then the change introducing those "added" files needs to be "committed", etc...)

**Comment by BilldarBagdar** [comment]
It seems to me that I only vaguely answered your first two questions above.  If you would like me to go ahead and do the "git work" I can.  But it needs to get done.  You have a pending Pull request to the "old master" branch without "folders", and now I have a new Pull request with the changes that (I think) make the script work in both Live 10 and Live 11. (Well, the actual changes are only in the 'legacy10support' branch, but that branch is in the "folders" repo configuration.)

We are not too far off the rails yet, but "origin/master" needs to always be the "source of truth".  And we need to maintain and care for "the master" with utmost attention to detail.  I won't issue a Pull request from "legacy support" to "master" until you/we get "master" all squared away.  

However, you can still review the minor changes I had to make for legacy support.  The only "non import" change was because you added the "umlaut vowels" to EncoderController and (I think) all I did to fix the problem was change the "file encoding" from ASCII to UTF-8 for that one file.  

Live 10 loads the script, yay! (but I don't have your latest "bug fix" because it's not in 'origin/master' yet, boo!). LOL :)

**Reply by markusschloesser** [reply]
> It seems to me that I only vaguely answered your first two questions above. If you would like me to go ahead and do the "git work" I can. But it needs to get done. You have a pending Pull request to the "old master" branch without "folders", and now I have a new Pull request with the changes that (I think) make the script work in both Live 10 and Live 11. (Well, the actual changes are only in the 'legacy10support' branch, but that branch is in the "folders" repo configuration.)

MS: Yes, if you could do that (folders), that'll be great! :-)
Do I need to "pull" ALL branches after that?


> We are not too far off the rails yet, but "origin/master" needs to always be the "source of truth". And we need to maintain and care for "the master" with utmost attention to detail. I won't issue a Pull request from "legacy support" to "master" until you/we get "master" all squared away.
> 
> However, you can still review the minor changes I had to make for legacy support. The only "non import" change was because you added the "umlaut vowels" to EncoderController and (I think) all I did to fix the problem was change the "file encoding" from ASCII to UTF-8 for that one file.
> 
> Live 10 loads the script, yay! (but I don't have your latest "bug fix" because it's not in 'origin/master' yet, boo!). LOL :)

MS: yay! :-) and boo! (you just need to change 1 line! 😉😂)

**Reply by BilldarBagdar** [reply]
Ok  - 

I will 
- Reject the Pull request to master that you assigned to me.  
- pull my two branches into master 
- and then delete them.
Then you will need to
- save aside your updated files away from git (or remember the one line change :))
- `git checkout master`
- `git pull origin master`
// This should give you all of my updates in ./wip/MackieC4/*.py files from origin/master in your local master branch.
// and none of "your" *.py files at the repo root
Then you need to decide how to proceed.  You can checkout a fresh branch and just move forward, you can Pull/merge master into your "stable" branch, you can delete the existing stable branch and start a new one,,. It's up to you.  I didn't see too many substantial changes in your Pull request.  I would suggest checking out a new (or fresh) stable branch from the fresh master and then copy/pasting your changes in from the backup files. (and committing them and pushing them, and then that Pull request should merge automatically into origin/master  (Then "remember" to Pull origin/master down to local/master :))

You never "need to" pull updated branches generally, except master.  You always want to stay current with master.  Maybe if there were dozen of contributors, one might only "Pull from origin/master" once a week or something.  But for this project, and just the two of us, whenever you know something was Pulled into origin/master, then Pull those updates down to local and merge with what you are doing.  Most of the time the pulls and merges should go smoothly.  But as soon as one doesn't, we'll be glad we practiced a little discipline.

**Reply by BilldarBagdar** [reply]
Done at origin.  Now I will update my local master from origin and delete the local branches that are "behind origin master" now.

**Reply by markusschloesser** [reply]
actually there was a little bit more then one line ;-)
Thanks so much for doing that!
I did my part, BUT: Is the WIP the master/main? So do I merge my stuff into this?

**Reply by markusschloesser** [reply]
@BilldarBagdar I thought I had understood it, but now, after importing the stuff from the last days into the "new main" branch, pycharm wants to do this:
![image](https://user-images.githubusercontent.com/59286549/112068540-68719300-8b6a-11eb-9570-640fe471afd5.png)
And I don't know why 🤔
I obviously did not push (for now).

**Comment by BilldarBagdar** [comment]
So, the folders (the project "folder structure") are the master branch now.  Tomorrow, we could change our mind and "remove" the folders, and that configuration would also be in "master".

But the repo still has the history from before the folders and it looks like PyCharm is remembering some of that history from I guess, when your branch was created.  Again, an IDE like PyCharm is not the best tool for managing repositories.  Strong integration exists and you can do a lot of stuff right in the IDE, and that's great.  But it only goes so far, and I'm afraid this is one of those situations where "recovery" from inside PtCharm using only the git tools PyCharm provides might not be possible.  I'm sure I could "fix the repo" using command line commands.

It looks like PyCharm wants to "restore" the three files you edited to their "old location" at root, and also save updates to their "new home".  But it also looks like maybe your local/master branch is not up to date with origin/master which is why the "restore" stuff.

If all you did was "rename" your local "master" branch to MS_stable, under the hood git or PyCharm might think of it as master still. (notice in the top middle where is shows master --> origin/master?  It's like Pycharm wants to merge you stable branch into your local master and push your local master to origin/master all in one fell swoop.  In my case, when I was doing all of the repo work yesterday, Pycharm just "responded and reloaded" every branch I checked out.  Meaning folders or not, the project would just reconfigure itself to however the "checked out branch" was organized (as soon as I "touched" the app window to bring it to focus).

Can you go to cmd.exe (or git-bash, or whatever) cd to your repo directory, and then run 'git status' and 'git branch -a' and share the results?
![image](https://user-images.githubusercontent.com/80170376/112073904-fe91c300-8b42-11eb-84e3-ebd1b77753d7.png)
![image](https://user-images.githubusercontent.com/80170376/112073949-1ec18200-8b43-11eb-8991-43fd747c2962.png)

You could also try Pulling origin/master into your local master (again?) and then start a new Branch directly from the "folder master" then you can copy/paste changes into that branch.  (save aside your changes or they will "hide" with their branch when you check out another branch.

**Reply by markusschloesser** [reply]
```
C:\Users\Markus\PycharmProjects\C4P3>git status
On branch master
Your branch is ahead of 'origin/master' by 6 commits.
  (use "git push" to publish your local commits)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        new file:   z_utility/__init__.py
        new file:   z_utility/tools MS/AbletonLive-API-Stub-master.zip
        new file:   z_utility/tools MS/AbletonLive-API-Stub-master/Live old.xml
        new file:   z_utility/tools MS/AbletonLive-API-Stub-master/Live.css
        new file:   z_utility/tools MS/AbletonLive-API-Stub-master/Live.xml
        new file:   z_utility/tools MS/AbletonLive-API-Stub-master/Live/__init__.py
        new file:   z_utility/tools MS/AbletonLive-API-Stub-master/README.md
        new file:   z_utility/tools MS/AbletonLive-API-Stub-master/generate-stub.py
        new file:   z_utility/tools MS/Live.css
        new file:   z_utility/tools MS/Live10.1.19.xml
        new file:   z_utility/tools MS/Live11.0.xml
        new file:   z_utility/tools MS/Live9.2.xml
        new file:   z_utility/tools MS/Live9.6.xml
        new file:   z_utility/tools MS/New Text Document.txt
        new file:   z_utility/tools MS/another uncompile.py
        new file:   z_utility/tools MS/bla2to3.py
        new file:   z_utility/tools MS/code_snippets_NOT_WORKING.py
        new file:   z_utility/tools MS/uncompy.py

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   wip/MackieC4/EncoderController.py
        modified:   wip/MackieC4/MackieC4.py

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        old/
```

and
```
C:\Users\Markus\PycharmProjects\C4P3>git branch -a
* master
  remotes/origin/MS_stable_test_for_while_devpm
  remotes/origin/master
```

But I already worked a bit on stuff again (and fixed the "Sends" :-) )

**Reply by markusschloesser** [reply]
I also installed sourcetree, looks great but I need to familiarize myself with it first

**Reply by BilldarBagdar** [reply]
Ok - Yes.  PyCharm has you on git's local master branch and git doesn't show your local "MS_stable" branch at all, it only shows "MS_stable_test_..." at origin.  The stuff under z_utility has been "staged for commit".  But, are you sure you want that `__init__.py` committed under z_utility?  The two files you modified still need to be "staged/added".  But otherwise, this doesn't really show me what PyCharm was showing, so I think the repo is "safe".  

Poke around in SourceTree.  If you want, you could clone a whole new fresh copy of the repo and just "play".  Nothing is even on the radar until something gets pushed up to origin.

**Reply by markusschloesser** [reply]
managed to get things done with sourcetree :-)
